industrial_taxonomy/pipeline/NSPL/flow.py

```diff
@@ -0,0 +1,158 @@
+
+    @step
+    def start(self):
+        """Get dynamically rendered download link."""
+        from utils import (
+            chrome_driver,
+            find_download_url,
+        )
+
+        self.geoportal_url = f"{GEOPORTAL_URL_PREFIX}/{self.geoportal_dataset}"
+        with chrome_driver() as driver:
+            self.download_url = find_download_url(driver, self.geoportal_url)
+
+        self.next(self.get_data)
+
+    @step
+    def get_data(self):
+        """Download zipped NSPL collection, extracting main lookup & LAUA names."""
+        import logging
+
+        import requests_cache
+
+        from utils import (
+            download_zip,
+            filter_nspl_data,
+            read_nspl_data,
+            read_laua_names,
+            nspl_joined,
+        )
+
+        if self.test_mode and not current.is_production:
+            nrows = 1_000
+            logging.warning(f"TEST MODE: Constraining to first {nrows} rows!")
+        else:
+            nrows = None
+
+        if not current.is_production:
+            requests_cache.install_cache("nspl_zip_cache")
+        with download_zip(self.download_url) as zipfile:
+            # Load main postcode lookup
+            self.nspl_data = read_nspl_data(zipfile, nrows).pipe(filter_nspl_data)
+
+            # LAUA lookup from codes to names
+            laua_names_tmp = read_laua_names(zipfile)
+
+            self.laua_names = laua_names_tmp
+            self.nspl_linked = nspl_joined(self.nspl_data, self.laua_names)
+            print(self.nspl_linked)
+            print(self.laua_names)
+            print(self.nspl_data.columns)
+
+        self.next(self.data_quality)
+
+    @step
+    def data_quality(self):
+        """Data quality checks."""
+        from utils import LOOSE_UK_BOUNDS
+
+        print(type(self.nspl_data))
+
+        # Null checks
+        has_nulls = self.nspl_data.isnull().sum().sum() > 0
+        if has_nulls:
+            raise AssertionError("Nulls detected")
+
+        # Postcode validity
+        # Choose very simple postcode verification as NSPL is a fairly
+        # authoritative source that may update faster than a precise regex
+        POSTCODE_REGEX = r"^([A-Z]{1,2}[A-Z\d]{0,2}? ?\d[A-Z]{2})$"
+        valid_pcds = self.nspl_data.index.str.match(POSTCODE_REGEX)
+        if not valid_pcds.all():
+            raise AssertionError(
+                "Invalid postcodes detected: "
+                f"{self.nspl_data.loc[~valid_pcds].index.values}"
+            )
+
+        # Check we have names for all laua codes
+        nspl_laua_cds = set(self.nspl_data.laua_code.dropna())
+        laua_names_cds = set(self.laua_names.keys())
+        laua_diff = nspl_laua_cds - laua_names_cds
+        # if len(laua_diff) > 0:
+        #   raise AssertionError(f"LAUA do not match: {laua_diff}")
+        #######lines 147 and 148 caused an error which state that the rows arent equal however would this be solved with the join? ##
```

Comment:
I mean lines 144 and 145.

These lines have caused an error raising an assertion error.